server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java

```diff
@@ -171,4 +250,82 @@ public Object terminatePartial(RamAccounting ramAccounting, CompareBy state) {
     public DataType<?> partialType() {
         return partialType;
     }
+
+
+    static class CmpByLongState {
+
+        static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(CmpByLongState.class);
+
+        long cmpValue = Long.MIN_VALUE;
+
+        int docId;
+        LeafReaderContext leafReaderContext;
+        boolean hasValue = false;
+    }
+
+    static class CmpByLong implements DocValueAggregator<CmpByLongState> {
+
+        private final int mod;
+        private final String columnName;
+        private final LuceneCollectorExpression<?> resultExpression;
+        private final DataType<?> searchType;
+
+        private SortedNumericDocValues values;
+        private LeafReaderContext leafReaderContext;
+
+        public CmpByLong(int mod,
+                         String columnName,
+                         DataType<?> searchType,
+                         LuceneCollectorExpression<?> resultExpression) {
+            this.mod = mod;
+            this.columnName = columnName;
+            this.searchType = searchType;
+            this.resultExpression = resultExpression;
+            resultExpression.startCollect(new CollectorContext());
+        }
+
+        @Override
+        public CmpByLongState initialState(RamAccounting ramAccounting,
+                                           MemoryManager memoryManager,
+                                           Version minNodeVersion) {
+            ramAccounting.addBytes(CmpByLongState.SHALLOW_SIZE);
+            return new CmpByLongState();
+        }
+
+        @Override
+        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
+            this.leafReaderContext = leafReaderContext;
+            values = DocValues.getSortedNumeric(leafReaderContext.reader(), columnName);
+        }
+
+        @Override
+        public void apply(RamAccounting ramAccounting, int doc, CmpByLongState state) throws IOException {
+            if (values.advanceExact(doc) && values.docValueCount() == 1) {
+                long value = values.nextValue();
+                if (value * mod > state.cmpValue * mod) {
```

Comment:
Long.MIN_VAL and MAX_VAL differ to 1 (by abs) which I think can lead to some overflow issues. Found at least one failing scenario, tested it locally

```
Signature signature = Signature.aggregate(
            "min_by",
            DataTypes.STRING.getTypeSignature(),
            DataTypes.LONG.getTypeSignature(),
            DataTypes.STRING.getTypeSignature()
        );
        Object result = executeAggregation(
            signature,
            new Object[][] {
                new Object[] { "a", 1l },
                new Object[] { "b", 2l },
                new Object[] { "c", Long.MIN_VALUE},
            },
            List.of()
        );
        assertThat(result).isEqualTo("c");
```
in debugger `partialResultWithoutDocValues` returns correct result "c"

not sure about other combinations - Long.MAX_VAL + (min/max by).   Long.Min_VALUE + `max_by` seems to be working.

Reply:
nice catch @BaurzhanSakhariev!